--- a/ext/libxml/libxml.c
+++ b/ext/libxml/libxml.c
@@ -41,6 +41,10 @@
 #endif
 
 #include "php_libxml.h"
+
+#ifndef ATTRIBUTE_UNUSED
+# define ATTRIBUTE_UNUSED __attribute__((unused))
+#endif
 #include "libxml_arginfo.h"
 
 #define PHP_LIBXML_ERROR 0
@@ -761,7 +765,11 @@
 #if LIBXML_VERSION >= 21200
 PHP_LIBXML_API void php_libxml_structured_error_handler(void *userData, const xmlError *error)
 #else
+#if LIBXML_VERSION >= 21200
+PHP_LIBXML_API void php_libxml_structured_error_handler(void *userData, const xmlError *error)
+#else
 PHP_LIBXML_API void php_libxml_structured_error_handler(void *userData, xmlErrorPtr error)
+#endif
 #endif
 {
 	_php_list_set_error_structure(error, NULL);
